### PR TITLE
Improve dataProcessing input/output handling

### PR DESCRIPTION
**Pull request by eparejatobes**

### CODE REVIEW

**Comment by eparejatobes** [comment]
@laughedelic please review. I left the previous state commented out.

**Comment by laughedelic** [comment]
ok. I'll review this a bit later today

File: src/main/scala/ohnosequences/loquat/dataProcessing.scala [hunk 1]
```diff
@@ -33,90 +36,71 @@ trait AnyProcessingContext {
...
+    Map[String, FileDataLocation]
+  ] { type Y = Either[ParseDenotationsError, FileLocationsFor[I]] }
 ) extends Bundle(deps: _*) with AnyDataProcessingBundle {
 
   type Input = I
-  // type InputFiles = I#Raw // with AnyKList.withBound[AnyDenotation { type Value = FileDataLocation }]
   type Output = O
-  // type OutputFiles =  O#Raw with AnyKList.withBound[AnyDenotation { type Value = FileDataLocation }]
+
+  // what's the point of instructions for this?
```
Comment (laughedelic) [inline src/main/scala/ohnosequences/loquat/dataProcessing.scala]: The point is that besides some bundle dependencies you may need to prepare each worker instance for the data processing and these instructions will be ran only once (before any processing).

File: src/main/scala/ohnosequences/loquat/dataMappings.scala [hunk 2]
```diff
@@ -27,13 +27,11 @@ trait AnyDataMapping {
...
-  // val outputsToMap: ToMap[RemoteOutput, AnyData, S3DataLocation]
-  def inputsToMap: Map[String,AnyS3Address] =
-    (dataProcessing.input.keys.types.asList map { t => t.label }) zip (remoteInput.asList map { d => d.value.location }) toMap
+  def inputsToMap: Map[String,AnyS3Address] = ???
+    // (dataProcessing.input.keys.types.asList map { t => t.label }) zip (remoteInput.asList map { d => d.value.location }) toMap
 
-  def outputsToMap: Map[String,AnyS3Address] =
-    (dataProcessing.output.keys.types.asList map { t => t.label }) zip (remoteOutput.asList map { d => d.value.location }) toMap
+  def outputsToMap: Map[String,AnyS3Address] = ???
+    // (dataProcessing.output.keys.types.asList map { t => t.label }) zip (remoteOutput.asList map { d => d.value.location }) toMap
```
Comment (laughedelic) [inline src/main/scala/ohnosequences/loquat/dataMappings.scala]: So in the end we need the input/output values for dataMapping serialization.

**Comment by laughedelic** [comment]
as I see the solution: 
- you can parametrize you data (sample stuff)
- then dataProcessingBundle is parametrized by the same thing
- you provide the input/output datasets values on construction of the dataProcessingBundle
- then inside you have these values and everybody's happy, the only implicit thing you need is the parser from `Map[String, File]` to `InputFiles` (which is essentially `Input#Raw` with a bound)

something like:

```scala
  case class processingBundle(sample: Sample) extends DataProcessingBundle()(
    input = DataSet(reads1(sample) :×: reads2(sample) :×: |[AnyData])
    output = DataSet(stats :×: mergedReads(sample) :×: |[AnyData])
  ) { ... }
```

File: src/test/scala/ohnosequences/loquat/dataProcessing.scala [hunk 3]
```diff
@@ -8,20 +9,27 @@ import better.files._
...
 
-  case object instructs extends DataProcessingBundle()(
-    inputData,
-    outputData
+  case class inputData(sample: Sample) extends DataSet(reads1(sample) :×: reads2(sample) :×: |[AnyData])
+  case class outputData(sample: Sample) extends DataSet(stats :×: mergedReads(sample) :×: |[AnyData])
+
+  case class processingBundle(sample: Sample) extends DataProcessingBundle()(
+    inputData(sample), 
+    outputData(sample)
```
Comment (laughedelic) [inline src/test/scala/ohnosequences/loquat/dataProcessing.scala]: @eparejatobes please, take a look at this example code. It doesn't encounter an implicit for the parser..

File: src/test/scala/ohnosequences/loquat/dataProcessing.scala [hunk 4]
```diff
@@ -8,20 +9,25 @@ import better.files._
...
 
-  case object instructs extends DataProcessingBundle()(
-    inputData,
-    outputData
+  case class inputData(sample: Sample) extends DataSet(reads1(sample) :×: reads2(sample) :×: |[AnyData])
+  case class outputData(sample: Sample) extends DataSet(stats :×: mergedReads(sample) :×: |[AnyData])
+
+  case class processingBundle(sample: Sample) extends DataProcessingBundle()(
+    inputData(sample),
+    outputData(sample)
```
Comment (laughedelic) [inline src/test/scala/ohnosequences/loquat/dataProcessing.scala]: @eparejatobes please, take a look at this example code. It doesn't encounter an implicit for the parser..

File: src/main/scala/ohnosequences/loquat/dataProcessing.scala (hunk 1 above)
Comment (eparejatobes) [inline src/main/scala/ohnosequences/loquat/dataProcessing.scala]: Fine. But I think it more reasonable to separate that and write a bundle doing it

**Comment by eparejatobes** [comment]
I already published in ohnosequences/cosas#107, which means we have access to type _values_ in denotations

**Comment by laughedelic** [comment]
ok. I'll use it here

**Comment by laughedelic** [comment]
In the end it's not about parametric input/output, but just simplifications and improvements taking into account changes in cosas (denotation now has a value of the `Tpe` type)

**Pull request merged**